src/config.rs

```diff
@@ -0,0 +1,134 @@
+
+/// The possible errors which the parser might emit.
+pub type ParserError = toml::de::Error;
+
+/// The configuration toplevl structure.
+#[derive(Clone, Debug, PartialEq, serde::Deserialize, serde::Serialize)]
+#[serde(tag = "version")]
+pub enum Config {
+    /// The first (and current) version of the config format.
+    #[serde(rename = "1")]
+    V1 {
+        /// Input parameter for the kernel and application
+        input: Input,
+
+        /// Minimal requirements for an image to be able to run as expected
+        #[serde(default)]
+        requirements: Requirements,
+
+        /// Kernel ELF file path
+        kernel: String,
+    },
+}
+
+/// Input parameter for the kernel and application
+#[derive(Clone, Debug, PartialEq, serde::Deserialize, serde::Serialize)]
+pub struct Input {
+    /// Arguments to be passed to the kernel
+    pub kernel_args: Vec<String>,
+
+    /// Arguments to be passed to the application
+    pub app_args: Vec<String>,
+
+    /// Environment variables
+    pub env_vars: Vec<String>,
+}
+
+/// Minimal requirements for an image to be able to run as expected
+#[derive(Clone, Debug, Default, PartialEq, serde::Deserialize, serde::Serialize)]
+pub struct Requirements {
+    /// Minimum RAM
+    pub memory: Option<byte_unit::Byte>,
+
+    /// Minimum amount of CPUs
+    #[serde(default)]
+    pub cpus: u32,
+}
+
+/// Parse a config file from a byte slice.
+#[inline]
+pub fn parse(data: &[u8]) -> Result<Config, ParserError> {
+    toml::from_slice(data)
+}
+
+/// An error from [`handle_config`].
+#[derive(Clone, Debug)]
+pub enum HandleConfigError {
+    /// The Hermit image couldn't be parsed.
+    TarParser(tar_parser::ParserError),
+
+    /// [`DEFAULT_CONFIG_NAME`]
+    /// either couldn't be found in the image or isn't a regular file.
+    ConfigResolve,
+    /// The Hermit image configuration file failed to parse.
+    ConfigParserError(ParserError),
+    /// The Kernel specified in the image configuration file
+    /// either couldn't be found in the image or isn't a regular file.
+    KernelResolve,
+}
+
+impl fmt::Display for HandleConfigError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::TarParser(p) => {
+                write!(f, "couldn't parse Hermit image: {p}")
+            }
+            Self::ConfigResolve => {
+                write!(f, "couldn't find Hermit image configuration file")
+            }
+            Self::ConfigParserError(e) => write!(f, "Hermit image configuration is invalid: {e}"),
+            Self::KernelResolve => write!(f, "couldn't find Hermit kernel in image"),
+        }
+    }
+}
+
+#[cfg(feature = "std")]
+impl std::error::Error for HandleConfigError {}
+
+/// A convenience function to handle looking up the config
+/// in a tar file (decompressed) and retrieve the kernel slice.
+pub fn handle_config(image: &tar_parser::Bytes) -> Result<(Config, &[u8]), HandleConfigError> {
+    fn lookup_in_image<'a>(
```

Comment:
This way of implementing this is a trade-off, of course, it reduces memory requirements but requires us to iterate over the tar file at least 3 times during boot instead of 2 times.